ql/src/java/org/apache/hadoop/hive/ql/exec/util/Retryable.java

```diff
@@ -0,0 +1,212 @@
+import org.apache.hadoop.hive.metastore.utils.SecurityUtils;
+import org.apache.hadoop.security.UserGroupInformation;
+
+import java.security.PrivilegedExceptionAction;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class to implement any retry logic in case of exceptions.
+ */
+public class Retryable {
+  private static long MINIMUM_DELAY_IN_SEC = 1;
+
+  private long totalDurationInSeconds;
+  private List<Class<? extends Exception>> retryOn;
+  private List<Class<? extends Exception>> failOn;
+  private long initialDelayInSeconds;
+  private long maxRetryDelayInSeconds;
+  private double backOff;
+  private int maxJitterInSeconds;
+
+  private Retryable() {
+    this.retryOn = new ArrayList<>();
+    this.failOn = new ArrayList<>();
+    this.initialDelayInSeconds = HiveConf.toTime(HiveConf.ConfVars.REPL_RETRY_INTIAL_DELAY.defaultStrVal,
+      HiveConf.getDefaultTimeUnit(HiveConf.ConfVars.REPL_RETRY_INTIAL_DELAY), TimeUnit.SECONDS);
+    this.maxRetryDelayInSeconds = HiveConf.toTime(HiveConf.ConfVars.REPL_RETRY_MAX_DELAY_BETWEEN_RETRIES.defaultStrVal,
+      HiveConf.getDefaultTimeUnit(HiveConf.ConfVars.REPL_RETRY_MAX_DELAY_BETWEEN_RETRIES), TimeUnit.SECONDS);
+    this.backOff = HiveConf.ConfVars.REPL_RETRY_BACKOFF_COEFFICIENT.defaultFloatVal;
+    this.maxJitterInSeconds = (int) HiveConf.toTime(HiveConf.ConfVars.REPL_RETRY_JITTER.defaultStrVal,
+      HiveConf.getDefaultTimeUnit(HiveConf.ConfVars.REPL_RETRY_JITTER), TimeUnit.SECONDS);
+    this.totalDurationInSeconds = HiveConf.toTime(HiveConf.ConfVars.REPL_RETRY_TOTAL_DURATION.defaultStrVal,
+      HiveConf.getDefaultTimeUnit(HiveConf.ConfVars.REPL_RETRY_TOTAL_DURATION), TimeUnit.SECONDS);;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public <T> T executeCallable(Callable<T> callable) throws Exception {
+    long startTime = System.currentTimeMillis();
+    long delay = this.initialDelayInSeconds;
+    Exception currentCapturedException = null;
+    while(true) {
+      try {
+        if (UserGroupInformation.isSecurityEnabled()) {
+          SecurityUtils.reloginExpiringKeytabUser();
+          return UserGroupInformation.getLoginUser().doAs((PrivilegedExceptionAction<T>) () -> callable.call());
+        } else {
+          return callable.call();
+        }
+      } catch (Exception e) {
+        if (this.failOn.stream().noneMatch(k -> e.getClass().equals(k))
+          && this.retryOn.stream().anyMatch(k -> e.getClass().isAssignableFrom(k))) {
+          if (elapsedTimeInSeconds(startTime) + delay > this.totalDurationInSeconds) {
+            // case where waiting would go beyond max duration. So throw exception and return
+            throw e;
+          }
+          sleep(delay);
+          //retry case. compute next sleep time
+          delay = getNextDelay(delay, currentCapturedException, e);
```

Comment:
nit: ( currentCapturedException and e) counterparts are (previousException, currentException) in method signature getNextDelay(). Consider changing the name, it looks a bit confusing